### ISSUE TITLE
Reproducibility problem ...

### ISSUE THREAD

**Issue by uturuncoglu** [uturuncoglu/RegESM#13]
TICKET: #3613762: ESMF and reproducibility ...

I am trying to test the reproducibility of the results in coupled model.,For this reason, i am using "-fp-model precise" compiler flag in all the model components and i tested them in standalone mode. I confirmed that the results of the individual components (i.e. atmosphere and wave) are reproducible. Then, i tested,coupled model but i could not reproduce the results. When i compare the,two different run with exactly same configuration, the results are,different. So, i just want to find the source of this issue and i would,like to narrow down the possible sources. My question is that do i need to,compile ESMF also with "-fp-model precise" compiler flag? Is there any,ESMF test to check the reproducibility in ESMF side? What do you suggest,about it? You could find the detail of my work environment as follows

ifort (IFORT) 14.0.1 20131008
openmpi-1.6.5 (compiled with same compiler)
esmf-7.0.0 (beta 38)
CentOS release 6.5 (Final) 2.6.32-431.29.2.el6.x86_64
Intel(R) Xeon(R) CPU E5-2660 0 @ 2.20GHz

**Comment by uturuncoglu** [comment]
After my initial mail, i read ESMF documentation and i decided to add srcTermProcessing (= 0) to ESMF_FieldRegridStore calls and termorderflag (= ESMF_TERMORDER_SRCSEQ) to ESMF_FieldRegrid calls to ensure bit-to-bit reproducibility. Then, i did two different simulation (along with exactly same input and configuration) for relatively short time period (one month). The results show that there are still difference between simulations when i compare the results.

**Comment by uturuncoglu** [comment]
Response from ESMF Team:

Those are the correct flags to make the ESMF sparse matrix multiply part bit-for-bit reproducible. However, I wonder what regrid scheme you are using? Is it conservative or bilinear or patch? Did you check whether the weights are bit-for-bit between runs? Just curious. I don't think
we expect issues with the weights, but also not sure we are 100% on that yet from our side. CC'ing Bob to chime in.

-Gerhard 

---
Just to let you know. I’ve made some efforts as I’ve implemented stuff to avoid causing problems with bit-for-bit, but we’ve never gone through and checked everything to verify that they’re actually bit-for-bit. We could certianly make that a feature request if you need it. 

- Bob

**Comment by uturuncoglu** [comment]
It seems that the returned factorLists are identical (in text files). I attached printed factor lists as reference. So, somehow i have still difference between runs. I also checked wave component again and its output is same in standalone version. So, it seems that driver is responsible from the difference. As i mention previously, do i need to compile ESMF with "-fp-model precise" option (for Intel compiler) to have a bit-to-bit reproducibility. In my case, standalone model components does not produce same results without this option?

**Comment by uturuncoglu** [comment]
Response from ESMF Team:

Okay, I agree that the next step would be to compile ESMF with the "-fp-model precise" option (for Intel compiler). It is kind of interesting that even without that option the factorList is identical if it is needed, but then again, it may just be how the numbers work out here.

You probably know this, but you can add extra compile flags like "-fp-model precise" by setting the ESMF_F90COMPILEOPTS and ESMF_CXXCOMPILEOPTS environment variables before building ESMF. 

Gerhard

**Comment by uturuncoglu** [comment]
I compared two different run compiled with ESMF 7.0 beta 38 (compiled with "-fp-model precise" for Intel compiler) and it seems that both runs are identical now. So, i could reproduce the results with this version of ESMF. In this case, i also compiled model component and the driver also with "-fp-model precise" flag. There might be a issue in sparse matrix multiplication algorithm (i.e. arithmetical operations) even weights are identical (i am not sure completely due to finite decimal representation
of factor lists in text files).

**Comment by uturuncoglu** [comment]
Response from ESMF Team:

We discussed this issue during the ticket meeting today. At first we were a bit puzzled by the fact that reproducibility of both your model, and of the ESMF sparse matrix multiplication would depend on the "-fp-model precise" compiler flag. Then Walter pointed out that the reductions (which definitely happen in the SMM, and maybe in your models, too) may be vectorized in the SIMD units of the processor, leading to partial sums. If the processor supports different order of summing these partial sums, this may introduce reproducibility issues.

I found this from Intel: 
https://software.intel.com/en-us/articles/consistency-of-floating-point-results-using-the-intel-compiler
You may be interested in this.

Bottom line is that auto-vectorization of reductions are turned OFF when setting the "-fp-model precise" flag. At this point I am assuming that the SMM bit-for-bit reproducibility can only be guaranteed if the compiler is instructed to NOT vectorize the reduction loops. 

Gerhard

**Comment by uturuncoglu** [comment]
The bit-to-bit reproducibility becomes important in the scientific studies when user needs to compare the results of the different configurations of the same model simulation to check the effects of the different components. The bit-to-bit reproducibility is still active research area in the earth system science and needs special attention to the used model components and the working environment (i.e. operating system, compiler). The issue is mainly related with the floating-point arithmetic and the representation of the numbers in the computer (it has a finite resolution to store the numbers). Due to this reason, numerical results change run to run of the same executable. Numerical results also changes between different systems. For example, using same operating system along with different CPU version (Intel® Xeon® Processor E5540 vs. Intel® Xeon® Processor E3-1275) probably affect the results. The problem could be caused by different reasons: the algorithm, alignment of the data in heap and stack (kind of memory), task or thread scheduler in OpenMP and MPI parallelism. The order of the numerical operations has also impact on the rounded results (i.e. reduction operator in MPI code).
The RegESM model is able to generate bit-to-bit reproducible results when specific compiler flags are used to compile/install the model components and also ESMF library. The detailed information about the the problem and the possible solution can be found in the Section 7.4 of User Guide.
